slice/auth.ice

```diff
@@ -3,11 +3,17 @@
 
 module Auth
 {
-    enum Role { User, Admin };
+    enum Role { User };
+    enum TokenVerificationStatus { Ok, InvalidToken, RoleNotAssigned};
+
+    struct AccessData {
+        string token;
+        Role role;
+    };
 
     interface Account
     {
-        bool verifyRole(Role role);
+        TokenVerificationStatus verifyToken(AccessData accessData);
```

Comment:
Imo User should be object containing userId(currently I can't identify user, I thought this would serve also as UserId provider) and user role. I understand the idea of inserting token and role, but I don't see any benefits coming from providing role as parameter. I believe it would be easier for microservices to acquire role, not insert it.

Reply:
Chodzi tu o to, że logika ról powinna być dla każdego mikroserwisu niezależna. W takiej formie w momencie dodania nowej roli, nie trzeba modyfikować wszystkich mikroserwisów, tylko te, na potrzebę których stworzono daną rolę. Przykład: mamy rolę user, którą używa 10 mikroserwisów. Chcemy dodać rolę user-premium, który ma np. możliwość dłuższych rezerwacji. W obecnym podejściu wystarczy lekko zmodyfikować serwis auth oraz rezerwacji, a reszta mikroserwisów nadal będzie działać. W przypadku, gdyby np. serwis wynajmu nagle dostał zamiast roli usera rolę user-premium, nie wiedziałby, co z tym zrobić. W praktyce przestałby poprawnie działać.

Reply:
poza tym, User zawiera UserID. Patrz: https://github.com/marszub/car-renting-system/blob/3c10c0a3ca8f842ee02e0f08ae9f2af2e3356bd4/auth/Auth-tmp/Models/User.cs#L5

Reply:
Dodatkowo, podejście pytania o rolę jest zgodne z ideą "zero knowledge proof". W tej sytuacji w kontekście bezpieczeństwa nie zmienia to wiele, ale implementuje tę zasadę w pełni, co jest w praktyce żadko spotykane i bardzo pożądane.